### ISSUE TITLE
[Bug] TS default parameters trigger react/require-default-props

### ISSUE THREAD

**Issue by Lesik** [iamturns/eslint-config-airbnb-typescript#203]
TypeScript allows specifying [default values for optional parameters](https://www.typescriptlang.org/docs/handbook/functions.html#optional-and-default-parameters) like so:

```ts
myFunction = (myParam1, myParam2 = "myValue2") => {};
```

This is the preferred way to specify default props for components, as shown in this [TypeScript example with React](https://www.typescriptlang.org/play/index.html?jsx=2&esModuleInterop=true&e=196#example/typescript-with-react) from the [official create-react-app docs](https://create-react-app.dev/docs/adding-typescript) from Facebook.

We are using `eslint-config-airbnb-typescript` exclusively and it complains about `react/require-default-props` on a component like this:

```tsx
interface MyProps {
  myParam1: string,
  myParam2?: string,
}

export default function MyComponent({
  myParam1,
  myParam2 = "myValue2",
}: MyProps) {
  return <></>;
}
```

One workaround that we found on accident is to export the props (`export interface MyProps`), even if they are not used anywhere, the eslint error goes away. I'm not sure if that's just because eslint stops parsing that line if it's exported (meaning it's just a hack to turn eslint off and basically equivalent to `eslint-disable-next-line`) or because it really should be exported.

**Comment by dahmc** [comment]
I just ended up using 
```
'react/require-default-props': 'off',
```

But I agree that this seems a bit off.

**Comment by vicasas** [comment]
The "problem" is not with this bookstore, but rather with `airbnb`. The default rule contains the `forbidDefaultForRequired` option which causes or forces default values ​​to be declared using `.defaultProps`.

Today luckily more options were added to the rule, where you can change the behavior for using default params using the `defaultArguments` option in `function`.

Its use would be:

```json
"react/require-default-props": [
    "error",
    {
        "functions": "defaultArguments"
    }
],
```

see: 

- https://github.com/airbnb/javascript/blob/11ab37144b7f846f04f64a29b5beb6e00d74e84b/packages/eslint-config-airbnb/rules/react.js#L392
- https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/require-default-props.md#functions

**Comment by SimonSimCity** [comment]
Yes, it is from `airbnb`, but I'd argue that `airbnb` is for a javascript context. Adding the typescript context, requiring this doesn't make sense (as I learned), because those props are mostly because you don't have types in javascript. But in typescript you should be able to rely on the types of your variables, which is why I think this rule should be disabled in the typescript package - which is this.

Anyways - this project hasn't seen many updates in a while, and I'm just going to disable it in my project config. Would be nice though to have this disabled in the config this repo provides.

**Comment by Kenneth-Sills** [comment]
I agree this would be the best approach, but it *does* deviate from the upstream configuration and @iamturns was really trying to make this 1:1.

Soon, I'll be working on the flat config for #331. With that, we can consider multiple different configurations and I think a new `recommended` configuration would be perfect for QoL improvements like this that *technically* aren't Airbnb compliant. Would that be alright with you, Matt?
